File: sdk/packages/eip8130-viem/src/address.ts
import {
  type Address,
  type Hex,
  concat,
  keccak256,
  toHex,
  hexToBytes,
  getAddress,
  slice,
} from "viem";
import type { Owner } from "./types.js";
import { DEPLOYMENT_HEADER_SIZE } from "./constants.js";

/**
 * Builds the 14-byte EVM deployment header that wraps bytecode for CREATE2.
 */
export function deploymentHeader(bytecodeLength: number): Hex {
  const hi = (bytecodeLength >> 8) & 0xff;
  const lo = bytecodeLength & 0xff;
  const bytes = new Uint8Array([
    0x61, hi, lo,
    0x80,
    0x60, 0x0e,
    0x60, 0x00,
    0x39,
    0x60, 0x00,
    0xf3,
    0x00, 0x00,
  ]);
  return toHex(bytes);
}

/**
 * Builds the full deployment code: header(len) || bytecode.
 */
export function deploymentCode(bytecode: Hex): Hex {
  const bytecodeBytes = hexToBytes(bytecode);
  const header = deploymentHeader(bytecodeBytes.length);
  return concat([header, bytecode]);
}

/**
 * Computes the effective_salt for CREATE2 address derivation.
 *
 * Owners are sorted by ownerId to make the address independent of ordering.
 * `effective_salt = keccak256(user_salt || keccak256(ownerId_0 || verifier_0 || scope_0 || ...))`
 */
export function effectiveSalt(userSalt: Hex, initialOwners: Owner[]): Hex {
  const sorted = [...initialOwners].sort((a, b) => {
    const aBytes = hexToBytes(a.ownerId as Hex);
    const bBytes = hexToBytes(b.ownerId as Hex);
    for (let i = 0; i < aBytes.length; i++) {
      if (aBytes[i] !== bBytes[i]) return aBytes[i] - bBytes[i];
    }
    return 0;
  });

  const commitmentParts: Hex[] = sorted.flatMap((owner) => [
    owner.ownerId as Hex,
    owner.verifier as Hex,
    toHex(owner.scope, { size: 1 }),
  ]);

  const ownersCommitment = keccak256(
    commitmentParts.length > 0 ? concat(commitmentParts) : "0x",
  );

  return keccak256(concat([userSalt, ownersCommitment]));
}

/**
 * Computes the CREATE2 address.
 *
 * `address = keccak256(0xff || deployer || salt || keccak256(initCode))[12:]`
 */
export function create2Address(
  deployer: Address,
  salt: Hex,
  initCode: Hex,
): Address {
  const initCodeHash = keccak256(initCode);
  const hash = keccak256(
    concat([
      "0xff",
      deployer as Hex,
      salt,
      initCodeHash,
    ]),
  );
  return getAddress(slice(hash, 12));
}

/**
 * Full address derivation for an account creation entry.
 */
export function deriveAccountAddress(
  deployer: Address,
  userSalt: Hex,
  bytecode: Hex,
  initialOwners: Owner[],
): Address {
  const salt = effectiveSalt(userSalt, initialOwners);
  const code = deploymentCode(bytecode);
  return create2Address(deployer, salt, code);
}

File: crates/common/consensus/src/transaction/eip8130/address.rs
//! EIP-8130 CREATE2 address derivation for account creation.

use alloc::vec::Vec;

use alloy_primitives::{Address, B256, Bytes, keccak256};

use super::types::Owner;

/// Builds the 14-byte EVM deployment header that wraps arbitrary bytecode for CREATE2.
///
/// The header is a minimal EVM program:
/// ```text
/// PUSH2 <len_hi> <len_lo>   // 0x61 LL HH
/// DUP1                       // 0x80
/// PUSH1 0x0e                 // 0x600e
/// PUSH1 0x00                 // 0x6000
/// CODECOPY                   // 0x39
/// PUSH1 0x00                 // 0x6000
/// RETURN                     // 0xf3
/// ```
pub fn deployment_header(bytecode_len: usize) -> [u8; 14] {
    let len = bytecode_len as u16;
    let hi = (len >> 8) as u8;
    let lo = (len & 0xFF) as u8;
    [
        0x61, hi, lo,   // PUSH2 len
        0x80, // DUP1
        0x60, 0x0e, // PUSH1 14  (header size = code offset)
        0x60, 0x00, // PUSH1 0
        0x39, // CODECOPY
        0x60, 0x00, // PUSH1 0
        0xf3, // RETURN
        0x00, 0x00, // padding to 14 bytes
    ]
}

/// Builds the full deployment code: `header(len) || bytecode`.
pub fn deployment_code(bytecode: &[u8]) -> Vec<u8> {
    let header = deployment_header(bytecode.len());
    let mut code = Vec::with_capacity(header.len() + bytecode.len());
    code.extend_from_slice(&header);
    code.extend_from_slice(bytecode);
    code
}

/// Computes the `effective_salt` for CREATE2 address derivation.
///
/// ```text
/// sorted_owners = sort(initial_owners, by: ownerId)
/// owners_commitment = keccak256(ownerId_0 || verifier_0 || scope_0 || ownerId_1 || ...)
/// effective_salt = keccak256(user_salt || owners_commitment)
/// ```
///
/// This makes the address independent of owner ordering.
pub fn effective_salt(user_salt: B256, initial_owners: &[Owner]) -> B256 {
    let mut sorted: Vec<&Owner> = initial_owners.iter().collect();
    sorted.sort_by_key(|o| o.owner_id);

    let mut commitment_input = Vec::with_capacity(sorted.len() * 53); // 32 + 20 + 1
    for owner in &sorted {
        commitment_input.extend_from_slice(owner.owner_id.as_slice());
        commitment_input.extend_from_slice(owner.verifier.as_slice());
        commitment_input.push(owner.scope);
    }
    let owners_commitment = keccak256(&commitment_input);

    let mut salt_input = Vec::with_capacity(64);
    salt_input.extend_from_slice(user_salt.as_slice());
    salt_input.extend_from_slice(owners_commitment.as_slice());
    keccak256(&salt_input)
}

/// Computes the CREATE2 address for an account deployment.
///
/// ```text
/// address = keccak256(0xff || deployer || effective_salt || keccak256(deployment_code))[12:]
/// ```
pub fn create2_address(deployer: Address, salt: B256, init_code: &[u8]) -> Address {
    let init_code_hash = keccak256(init_code);
    let mut buf = Vec::with_capacity(1 + 20 + 32 + 32);
    buf.push(0xFF);
    buf.extend_from_slice(deployer.as_slice());
    buf.extend_from_slice(salt.as_slice());
    buf.extend_from_slice(init_code_hash.as_slice());
    let hash = keccak256(&buf);
    Address::from_slice(&hash[12..])
}

/// Full address derivation for an account creation entry.
pub fn derive_account_address(
    deployer: Address,
    user_salt: B256,
    bytecode: &Bytes,
    initial_owners: &[Owner],
) -> Address {
    let salt = effective_salt(user_salt, initial_owners);
    let code = deployment_code(bytecode);
    create2_address(deployer, salt, &code)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deployment_header_is_14_bytes() {
        let h = deployment_header(256);
        assert_eq!(h.len(), 14);
        assert_eq!(h[0], 0x61); // PUSH2
        assert_eq!(h[1], 0x01); // 256 >> 8
        assert_eq!(h[2], 0x00); // 256 & 0xFF
    }

    #[test]
    fn deployment_code_concatenates() {
        let bytecode = [0x60, 0x00, 0xf3]; // PUSH1 0 RETURN
        let code = deployment_code(&bytecode);
        assert_eq!(code.len(), 14 + 3);
        assert_eq!(&code[14..], &bytecode);
    }

    #[test]
    fn effective_salt_order_independent() {
        let owner_a = Owner {
            verifier: Address::repeat_byte(1),
            owner_id: B256::repeat_byte(0x01),
            scope: 0,
        };
        let owner_b = Owner {
            verifier: Address::repeat_byte(2),
            owner_id: B256::repeat_byte(0x02),
            scope: 0,
        };

        let salt = B256::repeat_byte(0xAA);

        let s1 = effective_salt(salt, &[owner_a.clone(), owner_b.clone()]);
        let s2 = effective_salt(salt, &[owner_b, owner_a]);
        assert_eq!(s1, s2, "effective_salt must be independent of owner ordering");
    }

    #[test]
    fn create2_address_deterministic() {
        let deployer = Address::repeat_byte(0xDD);
        let salt = B256::repeat_byte(0xAA);
        let code = [0x60, 0x00, 0xf3];

        let a1 = create2_address(deployer, salt, &code);
        let a2 = create2_address(deployer, salt, &code);
        assert_eq!(a1, a2);
        assert_ne!(a1, Address::ZERO);
    }

    #[test]
    fn different_owners_different_address() {
        let deployer = Address::repeat_byte(0xDD);
        let salt = B256::repeat_byte(0xAA);
        let bytecode = Bytes::from_static(&[0x60, 0x00, 0xf3]);

        let owners_a = vec![Owner {
            verifier: Address::repeat_byte(1),
            owner_id: B256::repeat_byte(0x01),
            scope: 0,
        }];
        let owners_b = vec![Owner {
            verifier: Address::repeat_byte(2),
            owner_id: B256::repeat_byte(0x02),
            scope: 0,
        }];

        let a1 = derive_account_address(deployer, salt, &bytecode, &owners_a);
        let a2 = derive_account_address(deployer, salt, &bytecode, &owners_b);
        assert_ne!(a1, a2);
    }
}
